### PR TITLE
Add the `NoAccountAuthenticator` variant

### PR DESCRIPTION
**Pull request by junkil-park**
### Description
This PR updates the TypeScript SDK to support the simulation enhancement AIP: https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-92.md.

Key Changes:

Simulation API Update: Allows users to simulate transactions without providing public keys by updating simulateTransaction to accept `PublicKey | null` instead of `PublicKey`. If null is provided, `NoAccountAuthenticator` is used as an authenticator. This new authenticator variant is supported by the Aptos VM as per [this PR](https://github.com/aptos-labs/aptos-core/pull/13714).

Multisig V2 Example Update: The multisig v2 example (multisig_v2.ts) is updated to reflect the change in the multisig transaction simulation behavior. To pre-check the multisig payload before creation, an entry function payload simulation with the multisig account as the sender and 0x0 as the fee payer is used.

### Test Plan
This PR includes the following test scenarios:
* Simulate a transaction without checking the authentication key when the public key is not provided (i.e., public key is null).
* Simulate a fee payer transaction with fee payer as `0x0` and no public key of the fee payer provided.
* Fail to execute a transaction when `NoAccountAuthenticator` is used.

### Related Links
N.A.

### CODE REVIEW

File: tests/e2e/transaction/transactionSubmission.test.ts
Comment (0xmaayan) [inline tests/e2e/transaction/transactionSubmission.test.ts:845]: please follow the other test flows on this file, i.e

```
const transaction = await aptos.transaction.build.simple({...})
const authenticator = aptos.transaction.sign({...})
const response = await aptos.transaction.submit.simple({...})
```

File: src/transactions/transactionBuilder/transactionBuilder.ts
Comment (0xmaayan) [inline src/transactions/transactionBuilder/transactionBuilder.ts:452]: what are you trying to check here? It feels error-prone.

**Review by 0xmaayan — COMMENTED** [review]
left initial comments. can we gat a PR description on what we are changing/adding here and why?

**Comment by junkil-park** [comment]
> left initial comments. can we gat a PR description on what we are changing/adding here and why?

Added!

File: examples/typescript-esm/multisig_v2.ts
Comment (junkil-park) [reply examples/typescript-esm/multisig_v2.ts:142]: Got it. Removed. Thank you!

File: src/transactions/transactionBuilder/transactionBuilder.ts
Comment (junkil-park) [reply src/transactions/transactionBuilder/transactionBuilder.ts:452]: I updated this part so that `null` (rather than `000...0`) represents the absence of a public key.

File: src/cli/localNode.ts
Comment (junkil-park) [reply src/cli/localNode.ts:42]: This is a temporary change to run a custom built node locally. I will revert it before merging this PR later. I added "TODO" as a comment not to forget.

File: tests/e2e/transaction/transactionSubmission.test.ts
Comment (junkil-park) [reply tests/e2e/transaction/transactionSubmission.test.ts:845]: This test case may not fit into the usual test flow. It attempts to execute a transaction with `NoAccountAuthenticator` without signing. This should fail because `NoAccountAuthenticator` can be used only in simulation, not in execution. This case intentionally tries to misuse a lower level API such as `submitTransaction`.

File: src/transactions/transactionBuilder/transactionBuilder.ts
Comment (0xmaayan) [inline src/transactions/transactionBuilder/transactionBuilder.ts:450]:
```suggestion
export function getAuthenticatorForSimulation(publicKey?: PublicKey) {
```

File: src/transactions/transactionBuilder/transactionBuilder.ts
Comment (0xmaayan) [inline src/transactions/transactionBuilder/transactionBuilder.ts:451]:
```suggestion
  if (!publicKey) {
```

File: src/transactions/types.ts
Comment (0xmaayan) [inline src/transactions/types.ts:281]:
```suggestion
  signerPublicKey?: PublicKey;
```

File: src/transactions/types.ts
Comment (0xmaayan) [inline src/transactions/types.ts:285]:
```suggestion
  secondarySignersPublicKeys?: Array<PublicKey>;
```

File: src/transactions/types.ts
Comment (0xmaayan) [inline src/transactions/types.ts:289]:
```suggestion
  feePayerPublicKey?: PublicKey;
```

File: tests/e2e/transaction/transactionSubmission.test.ts
Comment (0xmaayan) [reply tests/e2e/transaction/transactionSubmission.test.ts:845]: I understand we dont need to sign, but we should still use `await aptos.transaction.submit.simple({...})` to submit the transaction

File: tests/e2e/transaction/transactionSimulation.test.ts
Comment (0xmaayan) [inline tests/e2e/transaction/transactionSimulation.test.ts:617]:
```suggestion
      const [response] = await aptos.transaction.simulate.simple({
```

File: tests/e2e/transaction/transactionSimulation.test.ts
Comment (0xmaayan) [inline tests/e2e/transaction/transactionSimulation.test.ts:652]: dont need it

File: tests/e2e/transaction/transactionSimulation.test.ts
Comment (0xmaayan) [inline tests/e2e/transaction/transactionSimulation.test.ts:657]:
```suggestion
      transaction: rawTxn,
```

File: tests/e2e/transaction/transactionSimulation.test.ts
Comment (0xmaayan) [inline tests/e2e/transaction/transactionSimulation.test.ts:639]:
```suggestion
      transaction: rawTxn,
```

File: tests/e2e/transaction/transactionSimulation.test.ts
Comment (0xmaayan) [inline tests/e2e/transaction/transactionSimulation.test.ts:676]:
```suggestion
      transaction: rawTxn,
```

**Review by 0xmaayan — COMMENTED** [review]
Did a first round
- Please fix the linting errors - can run `pnpm lint` on the sdk root folder
- When we have a property `publicKey?:PublicKey` it means the `publicKey` is of type `PublicKey` or `undefined` - please use this syntax
- Please update CHANGELOG
- Also, tests are failing

File: src/api/transactionSubmission/simulate.ts
Comment (junkil-park) [reply src/api/transactionSubmission/simulate.ts:35]: I think we need `null` (or something equivalent) to cover the new use case: 
`feePayerPublicKey` can have:
(1) `undefined` if it is not a fee-payer transaction,
(2) `somePublicKey` if it is a fee-payer transaction, and the public/auth key will be checked on-chain, or
(3) `null` if it is a fee-payer transaction, and the public/auth key check will be skipped on-chain.
Among these, (3) is the new use case.

File: src/api/transactionSubmission/simulate.ts
Comment (junkil-park) [reply src/api/transactionSubmission/simulate.ts:56]: I proposed `secondarySignersPublicKeys: Array<PublicKey | null>` because some secondary signers may provide public keys for auth key check, while others may not. For example, `secondarySignersPublicKeys` can be `[publicKey1, null, publicKey3, null]` where only first and third secondary signers provided public keys.

File: src/api/transactionSubmission/simulate.ts
Comment (0xmaayan) [reply src/api/transactionSubmission/simulate.ts:35]: I see. imo, it is a bit weird (in the SDK context and in the JS dev ecosystem) to differentiate between `null` and `undefined` and specifically ask the dev to write `feePayerPublicKey: null` as it means something critical.

Following this change - what would be the cases a dev still wants to NOT skip the public/auth key check? Are we (i.e Aptos) will be going with skipping the check from now on? wondering if we can combine the 1 and 3 use cases and just skip the check by default.

If we still want to keep the check, I would suggest maybe to introduce a new optional property for the dev to provide if it is a fee-payer transaction, and the public/auth key check will be skipped on-chain, something like
```
skipFeePayerPublicKeyCheck?:boolean
```

Comment (junkil-park) [reply src/api/transactionSubmission/simulate.ts:35]: > Following this change - what would be the cases a dev still wants to NOT skip the public/auth key check? Are we (i.e Aptos) will be going with skipping the check from now on? wondering if we can combine the 1 and 3 use cases and just skip the check by default.

Using a simulation, a dev may want to validate the public key that they have. Otherwise, they have to actually submit a transaction. (1) is different from (2) and (3) in terms of transaction types (fee-payer txn vs. simple txn).

Comment (junkil-park) [reply src/api/transactionSubmission/simulate.ts:35]: > If we still want to keep the check, I would suggest maybe to introduce a new optional property for the dev to provide if it is a fee-payer transaction, and the public/auth key check will be skipped on-chain, something like
> skipFeePayerPublicKeyCheck?:boolean

This sounds good for the fee-payer case!

However, it seems not clear how to apply in the secondary signer case: https://github.com/aptos-labs/aptos-ts-sdk/pull/467/files#r1686804774

Comment (0xmaayan) [reply src/api/transactionSubmission/simulate.ts:35]: the secondary signer case makes sense to me - thanks for clarifying

**Review by 0xmaayan — COMMENTED** [review]
Let's address the PR comments, clean up the PR (lint and format errors) - then can do another final round

File: src/transactions/transactionBuilder/transactionBuilder.ts
Comment (junkil-park) [reply src/transactions/transactionBuilder/transactionBuilder.ts]: Update this.

File: src/api/transactionSubmission/simulate.ts
Comment (junkil-park) [reply src/api/transactionSubmission/simulate.ts]: Updated it accordingly.

File: src/api/transactionSubmission/simulate.ts
Comment (junkil-park) [reply src/api/transactionSubmission/simulate.ts]: Updated it according to our discussion.

File: src/api/transactionSubmission/simulate.ts
Comment (junkil-park) [reply src/api/transactionSubmission/simulate.ts]: Updated it accordingly.

File: src/api/transactionSubmission/simulate.ts
Comment (junkil-park) [reply src/api/transactionSubmission/simulate.ts]: Updated it to `secondarySignersPublicKeys: Array<PublicKey | undefined>;`

File: tests/e2e/transaction/transactionSimulation.test.ts
Comment (junkil-park) [reply tests/e2e/transaction/transactionSimulation.test.ts:649]: Removed it.

File: tests/e2e/transaction/transactionSimulation.test.ts
Comment (junkil-park) [reply tests/e2e/transaction/transactionSimulation.test.ts]: Done updating it accordingly.

File: tests/e2e/transaction/transactionSimulation.test.ts
Comment (junkil-park) [reply tests/e2e/transaction/transactionSimulation.test.ts]: Updated it.

File: tests/e2e/transaction/transactionSimulation.test.ts
Comment (junkil-park) [reply tests/e2e/transaction/transactionSimulation.test.ts]: Updated it.

File: tests/e2e/transaction/transactionSimulation.test.ts
Comment (junkil-park) [reply tests/e2e/transaction/transactionSimulation.test.ts:617]: Done!

**Comment by junkil-park** [comment]
> Did a first round
> 
> * Please fix the linting errors - can run `pnpm lint` on the sdk root folder
Some are unavoidable, so I added `eslint-disable-next-line` for those cases.

> * When we have a property `publicKey?:PublicKey` it means the `publicKey` is of type `PublicKey` or `undefined` - please use this syntax
Done accordingly.

> * Please update CHANGELOG
Updated.

> * Also, tests are failing
They passed locally. It requires a custom localnet.

File: examples/typescript/local_node.ts
Comment (0xmaayan) [inline examples/typescript/local_node.ts:110]: 🙇‍♀️

File: src/api/transactionSubmission/simulate.ts
Comment (0xmaayan) [inline src/api/transactionSubmission/simulate.ts:46]: lets add some more info, something like
```suggestion
   * @param args.secondarySignersPublicKeys An array of the secondary signers public keys, each of which can be optional to support the public key skip checking
```

File: src/transactions/authenticator/account.ts
Comment (0xmaayan) [inline src/transactions/authenticator/account.ts:176]: could we elaborate when/why one would want to use it?

File: src/transactions/types.ts
Comment (0xmaayan) [inline src/transactions/types.ts:283]: please add some more explanation on why/when a public key can be undefined

File: tests/e2e/transaction/transactionSubmission.test.ts
Comment (0xmaayan) [inline tests/e2e/transaction/transactionSubmission.test.ts:833]: can we add more tests for all the different public key use cases? with a pk, without a pk, with array of pks (some defined some arent), etc

**Review by 0xmaayan — COMMENTED** [review]
whats the migration process? looks like we first need to have the feature on mainnet? or are we good with only on devnet/testnet?
Can we make sure tests are passing with the localnet on main branch before merging it in?

File: tests/e2e/transaction/transactionSimulation.test.ts
Comment (0xmaayan) [inline tests/e2e/transaction/transactionSimulation.test.ts:606]: can we add more tests for all the different public key use cases? with a pk, without a pk, with array of pks (some defined some arent), etc

File: tests/e2e/transaction/transactionSubmission.test.ts
Comment (0xmaayan) [inline tests/e2e/transaction/transactionSubmission.test.ts:834]:
```suggestion
    test("it fails to submit a transaction when authenticator in undefined", async () => {
```

**Review by movekevin — APPROVED** [review]

File: tests/e2e/transaction/transactionSubmission.test.ts
Comment (junkil-park) [reply tests/e2e/transaction/transactionSubmission.test.ts]: Updated it.

File: src/transactions/types.ts
Comment (junkil-park) [reply src/transactions/types.ts:284]: Added some explanation.

File: src/api/transactionSubmission/simulate.ts
Comment (junkil-park) [reply src/api/transactionSubmission/simulate.ts]: Added some info

File: tests/e2e/transaction/transactionSimulation.test.ts
Comment (junkil-park) [reply tests/e2e/transaction/transactionSimulation.test.ts:606]: I add the test cases for the multiagent transaction cases (for array of pks) : https://github.com/aptos-labs/aptos-ts-sdk/pull/467/files#diff-f97493b76e508e849a876ecb35492119c3ebf540140bda3c1a7b3313a32be2deR670-R708

File: examples/typescript-esm/multisig_v2.ts
Comment (0xmaayan) [inline examples/typescript-esm/multisig_v2.ts:145]: can be removed. it is on devnet https://aptos-org.slack.com/archives/C05NLAKJM9U/p1726612173049729

File: src/transactions/transactionBuilder/transactionBuilder.ts
Comment (0xmaayan) [inline src/transactions/transactionBuilder/transactionBuilder.ts:427]:
```suggestion
      secondaryAccountAuthenticators = secondarySignersPublicKeys.map((publicKey) =>
```

File: src/transactions/transactionBuilder/transactionBuilder.ts
Comment (0xmaayan) [inline src/transactions/transactionBuilder/transactionBuilder.ts:426]: repeated code, lets move it into a function

**Comment by gregnazario** [comment]
I rebased and fixed the conflicts.

**Comment by junkil-park** [comment]
> I rebased and fixed the conflicts.

Thank you! We’re now waiting on the security review and the testnet/mainnet release before merging this PR.

File: examples/typescript-esm/multisig_v2.ts
Comment (junkil-park) [reply examples/typescript-esm/multisig_v2.ts]: Done removing it.

File: src/transactions/transactionBuilder/transactionBuilder.ts
Comment (junkil-park) [reply src/transactions/transactionBuilder/transactionBuilder.ts]: @0xmaayan , I updated the code, reducing the required step.

**Pull request merged**
